Image: Add border block support for color, width, and style

aaronrobertshaw opened the pull request:
Fixes: #38801

Depends on: 
- https://github.com/WordPress/gutenberg/pull/42087

Part of: 
- https://github.com/WordPress/gutenberg/issues/21540

## Description

This takes advantage of border block support to add those features to the Image block.

- Individual image border support added via block supports (skips serialization and manually applies to the inner img)
- Leverages block support feature level selectors added in [#42087](https://github.com/WordPress/gutenberg/pull/42087) to allow theme.json and Global Styles theming of borders.
- Theme.json can style the image borders via normal without having to use the additional nesting of the elements API

## How has this been tested?
Fixture tests and manually.

#### Note: There is a current bug on `trunk` impacting resetting of controls. 

#### Test Instructions
1. Create a post, add an image block, set a border-radius on the image block and save the post
2. Checkout this PR and reload the editor
3. Ensure there are no block validation errors and the block still displays its border-radius correctly
4. Set new border color, radius and width on the image block ensuring it updates correctly
5. From the block controls toolbar, choose to crop the image
6. Ensure the border still displays
7. Cancel the cropping and save the post
8. Confirm the image border displays correctly on the frontend
9. Back in the editor add a gallery, with images. Apply different borders and captions to the images.
10. After ensuring gallery images display ok, save the post and confirm for the frontend
11. Try styling the image block via theme.json (see snippet below)
12. Switch to the site editor and navigate within the Global Styles sidebar to Blocks > Image > Layout and test configuring borders on images.
12. Run `npm run test-unit` and ensure the image fixture tests pass

```
{
	"styles": {
		"core/image": {
			"border": {
				"width": "15px",
				"style": "dashed",
				"color": "blue"
			}
		}
	}
}
```

## Screenshots

https://user-images.githubusercontent.com/60436221/179129538-44627ad9-e892-4325-85db-45563e4a4916.mp4


## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Add border color, style and width support to image block

aaronrobertshaw committed:
Pass image border props to image cropper

This is skipped for the rounded block style.

aaronrobertshaw committed:
Add deprecation for moving border radius and new classname

New border block support styles and classes need to be on the inner img element. As a result it makes sense to move the border radius style with them. This requires a deprecation which this commit provides along with new fixtures.

.has-custom-border is required to handle images in galleries that have borders. This includes border radii which means we need to retroactively add it to existing image blocks that have that set.

paaljoachim commented:
Testing by downloading the special Gutenberg plugin. Procedure:
https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

I copied the TT1 experimental-theme.json file into Twenty Twenty One theme root folder. Renamed experimental-theme.json to theme.json, and replaced all the contents with:

```
"border": {
				"customColor": true,
				"customRadius": true,
				"customStyle": true,
				"customWidth": true
			}
```

Did a hard refresh. Even installed Autoptimize. Made another post. I have not been able to see the Border panel settings yet. I assume there is something I am not doing right.

aaronrobertshaw commented:
@paaljoachim thanks for testing this. It's working ok for me in the TwentyTwentyOne theme.

> Renamed experimental-theme.json to theme.json, and replaced all the contents with:

I think the problem might be the vague testing instructions. I've updated them to include a link to the current theme.json file I last tested with. 

If you did indeed replace all the `experimental-theme.json` contents with only the border config, the shape wouldn't be correct so the checks for whether the support is enabled and to show the inspector controls would fail.

[Example `experimental-theme.json` with border block support enabled.](https://gist.github.com/aaronrobertshaw/5284c32092d00f1215ffda8d9dba70ca#file-experimental-default-theme-json-L215)

aaronrobertshaw commented:
#### Core Block Styles
In general, there is only the "Rounded" block style by default. This style could be removed after a further iteration to the border block support to allow a greater max border radius value.

#### TwentyTwentyOne Block Styles
The TwentyTwentyOne and TT1 Blocks themes also introduce "Borders" and "Frame" block styles. These styles could be achieved via the border block support (and padding for "Frame" style). The block styles do however give the user a simple, one-click option to set a consistent border.

#### Summary 
If we were to remove the block styles, existing blocks with those styles would keep the `is-style-<style>` CSS class without the need for a deprecation. The class would appear under the Advanced CSS inspector control field.

Ultimately, I think whether we remove the block styles or not comes down to convenience and simplicity for the user at the price of some overlap and duplication between block styles and the block support features.

Happy to hear others' thoughts on this and/or iterate on the block styles in a separate PR.

vdwijngaert reviewed (approved):
Looks good as far as I can tell! Works as expected and unit tests pass. Might need some more work when it comes to the block styles part. It could be confusing to have the width and color controls there when the border style is set to `none` (or `default`, depending on defaults), though.

But, as you said, but that can probably happen later down the line.

aaronrobertshaw commented:
Thanks for testing and the feedback.

> Might need some more work when it comes to the block styles part. It could be confusing to have the width and color controls there when the border style is set to `none` (or `default`, depending on defaults), though.

I agree, this would be a nice improvement. 

One difficulty I've previously run into in this area is there isn't a reliable means to retrieve what the "default" value of any global style is in the normal post editor. In the site editor these are available within the state. As more block support features are adopted, I think the need to address this will increase.

Also, there is an open issue to refine the border block support controls https://github.com/WordPress/gutenberg/issues/31337. Part of the approach in that was to order the controls in line with the CSS order. Hiding and showing controls might not work well there.  Either way, the actual block support controls will need to be refined in a separate PR as suggested.

paaljoachim commented:
Hey Aaron. 

Thank you for working on this amazing PR! It is great to get Border settings in place!

Copied your experimental-theme.json file into Twenty Twenty One.

This is what I see. Brainstorming.

Border style drop down. Selecting "None" should hide the controls below.
I am wondering if "Default" and "None" both should have the Border controls hidden. (As already mentioned by @vdwijngaert ) Default could also be whatever Global styles have defined Image Border settings to be.

I noticed that when I move the slider a tooltip shows up, and will stick in place until I click somewhere else.

![Screen Shot 2021-05-06 at 00 52 29](https://user-images.githubusercontent.com/5323259/117219385-71aa7c80-ae05-11eb-9378-e4e38fa7e9d9.png)

I went through and checked each Border style type - Solid, Dashed and Dotted. Backend and frontend looked identical.

Image Styles. These do not work correctly when Border affects are in place. Such as selecting a dotted frame with a Border width of 40, and Border radius of 30. Example clicking Rounded did not show a rounded image.

![Screen Shot 2021-05-06 at 00 59 10](https://user-images.githubusercontent.com/5323259/117219805-4e340180-ae06-11eb-9bbb-18307a20ebef.png)

Clicking through Rounded (no affect), Borders (no affect) and Frame (made the image smaller).
This will gradually become a bigger issue for all blocks as more and more styling controls are added. Such as the Border, padding, margin and other controls. As specific Styles panel pre-created blocks will likely stop working as the user decides to instead use the specific custom controls a block offers. As there will be some clashes between these. There will likely come to a point where a user can style a block and save it as a style that is automatically added to the Styles panel. As it could then replace the default pre-created styles offered. (One would likely then also be able to click a pre-created style and see the various custom settings show the style through the controls.) Before that time there will just be some clashes between pre-created styles and the styles made by the user through the available controls. 

I turned off all adjustments I have made through Border settings also the border color and now the rounded style worked again. The Rounded style is a bit glitchy when experimenting with the Border settings.

aaronrobertshaw commented:
Great feedback @paaljoachim! Thanks for taking the time to be so thorough.

> I am wondering if "Default" and "None" both should have the Border controls hidden.

There is already an [open issue](https://github.com/WordPress/gutenberg/issues/31337) to refine the border support UI. It will actually do away with the `default` and `none` options. The plan there is to use the border width to control the border display. 

I'm actually in the process of working on the refinement of the border support UI. There are some blockers for it though which include:
- supporting non-pixel units for border width and radius
- implementing/adding new color and radius controls
- splitting the border-radius attribute so individual corner radii can be controlled

There is also a second PR [adding border support to the group block](https://github.com/WordPress/gutenberg/pull/31370). Both that and this PR won't land before the border block support UI refinements are complete. 

> I noticed that when I move the slider a tooltip shows up, and will stick in place until I click somewhere else.

As part of the refinements mentioned above, the both the width and radius RangeControls are being replaced so hopefully this won't be an issue moving forward.

> Image Styles. These do not work correctly when Border affects are in place. Such as selecting a dotted frame with a Border width of 40, and Border radius of 30. Example clicking Rounded did not show a rounded image.

Yes, the specificity of the border block support styles (inline styles) is greater than the CSS class styles applied via the block style e.g. Rounded.

In my [earlier comment on the block styles](https://github.com/WordPress/gutenberg/pull/31366#issuecomment-832308051) I listed some options. We could remove the block style entirely but the block styles do offer a simple one click configuration option for users. Another possibility is upon changing the selected block style we could clear the border support attributes. That might be the best compromise option.

> Frame (made the image smaller)

This is due to the Frame block style actually applying padding. Given the width of the block doesn't change its the image that shrinks in this case. All the border styles have been overridden via the deliberate selections in the border panel.

> There will likely come to a point where a user can style a block and save it as a style that is automatically added to the Styles panel.
> Before that time there will just be some clashes between pre-created styles and the styles made by the user through the available controls.

Agreed. Given themes can add their own block styles as well I'm not sure we can remove all possibility of conflict there. Having an agreed upon approach to handling it would certainly help. Especially given this will be the case for all block support provided styles not just borders.

aaronrobertshaw commented:
@jasmussen This PR has just been rebased to incorporate the recent border support UI changes. Do you feel this is in a place where we can move forward on adding border control to image blocks?

![ImageBordersRebased](https://user-images.githubusercontent.com/60436221/125038526-da012d00-e0d8-11eb-9ca5-7736cebb8f58.gif)

jasmussen commented:
This is what I see:

![image border](https://user-images.githubusercontent.com/1204802/125040567-1bbdb380-e098-11eb-9107-c47b9e7b7756.gif)

> This PR has just been rebased to incorporate the recent border support UI changes. Do you feel this is in a place where we can move forward on adding border control to image blocks?

I left a few comments on [the Cover PR which also apply here](https://github.com/WordPress/gutenberg/pull/31370#issuecomment-876973454). Things like setting a default border style and width if you start with a color, perhaps moving the border panel (if collapsed) to the bottom next to Advanced.

I also noticed that the "rounded" style no longer works for when a border radius is applied; I think this speaks to us needing to replace that style variation with a "preset shortcut". I.e. instead of applying `is-style-rounded`, now that border radius support is built in, clicking that thumbnail is just a shortcut to pre-configuring the border radius. This is likely a larger undertaking: style variations need a lot of love. But we need to think about what happens in the mean time, so both still work: which one takes precedence? Not entirely sure what the best interim solution here is, even if we know the longer term solution is a reimagining of the styles.

On a more high level, about all these border support PRs (thank you for your work!), here's my thinking about them:

- Because they leverage the border support design by Jay, we are in decent place with the usability (#31333 and #31337)
- If we can fix the small flow issues (see [comments on Cover PR](https://github.com/WordPress/gutenberg/pull/31370#issuecomment-876973454)), the flow will be addressed.
- We’ll want to, in the future (probably after the [new color swatch picker](https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866)) revisit where border support lives. Just like how we are looking at unifying width/height tools in a [unified Dimensions panel](https://github.com/WordPress/gutenberg/issues/28356), I suspect there's an opportunity to rethink appearance related tools. We might end up back on a dedicated Border (or Stroke?) panel, but it's something to think about.
- For this initial version, and especially since the panel is starting out collapsed, it seems like it's in a good place to launch for a few notable blocks. Probably not Paragraph, for example, but Pullquote and Cover seem good, probably even Group. This is considering the value these tools will bring to enabling patterns.

aaronrobertshaw commented:
Thanks for the great feedback as always Joen!

> Things like setting a default border style and width if you start with a color, perhaps moving the border panel (if collapsed) to the bottom next to Advanced.

I've replied in detail with [my thoughts](https://github.com/WordPress/gutenberg/pull/31370#issuecomment-877969470) on the suggestions on the Cover block PR. Hopefully, they are clear enough.

Essentially, those thoughts boil down to our block supports being overrides for the merged style value coming from theme.json, global styles etc. We don't have access to those merged values in the block editor at present. Try to set any default on the block support provided controls would likely mean blindly overriding theme values the use might be happy with already.

I do appreciate the point being made though regarding the flow.

At the moment the simpler case is how it currently works which is consistent across block supports. It could also be the option that requires the least re-editing of values by the user depending on their theme.

> I also noticed that the “rounded” style no longer works for when a border radius is applied; I think this speaks to us needing to replace that style variation with a “preset shortcut”.

Good observation. This has actually be spoken about and an [issue created](https://github.com/WordPress/gutenberg/issues/33232) to track the discussion. The idea is to explore block style variations as sets of block support style attributes.

As you mentioned, the rounded block style is effectively just a shortcut to applying a standard border radius to the image


> But we need to think about what happens in the mean time, so both still work: which one takes precedence? 

In technical terms, the border block support can only apply the border radius via inline styles. It will take precedence short of  `!important` being resorted to.

Conceptually, if a user is going to the lengths of expanding the border support panel and configuring an explicit border radius value, that makes sense to me in terms of taking precedence. 

A rounded style that is kept consistent across images also sounds perfectly reasonable. The block style does offer that consistency until such a time as we have implemented styles as sets of style attributes.

> On a more high level, about all these border support PRs (thank you for your work!), here’s my thinking about them:
> 
> If we can fix the small flow issues (see  [comments on Cover PR](https://github.com/WordPress/gutenberg/pull/31370#issuecomment-876973454) ), the flow will be addressed.

I’m not sure the suggested flow fixes work 100% with how block supports function. If we have to change how all block supports work, these will fast become large issues.

> We’ll want to, in the future (probably after the  [new color swatch picker](https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866) ) revisit where border support lives.

Agreed. That can be discussed and explored separately. We’ll be in a better place to do that once we have the new [panel](https://github.com/WordPress/gutenberg/pull/32392) in place.

> For this initial version, and especially since the panel is starting out collapsed, it seems like it’s in a good place to launch for a few notable blocks. Probably not Paragraph, for example, but Pullquote and Cover seem good, probably even Group.

Each block opting into the support will go require a PR to do so. That will provide the chance to evaluate them as required. 

By the way, the group block already has been opted into border support. It’s only a matter of whether the theme enables those controls or not.

Thanks again for all the considered responses. It definitely helps me work out what I'm missing to progress these PRs.

aaronrobertshaw committed:
Provide baseline border styles for improved visual feedback

aaronrobertshaw commented:
I've rebased this PR and added some baseline border styles in line with the discussion on the [Cover block border support PR](https://github.com/WordPress/gutenberg/pull/31370). This also required adding the same baseline styles for the image crop area.

mtias commented:
I'm curious why the properties are:

```json
"border": {
	"customColor": true,
	"customRadius": true,
	"customStyle": true,
	"customWidth": true
}
```

Instead of:

```json
"border": {
	"color": true,
	"radius": true,
	"style": true,
	"width": true
}
```

Given that there are no presets for radius, etc, so the custom prefix is spurious.

Also, let's disable border style for this one initially.

oandregal commented:
To the above comment by Matías: I think we've lost a bit the original meaning of the `custom*` prefix in `theme.json`. In origin, it was used to distinguish things that had both presets and the user could add a custom value. However, we've landed some things in WordPress 5.8 that used that prefix and could have done without it.

We should plan to increment the theme.json schema version in WordPress 5.9 and I wonder if we can do that change as well. Tracking it separately at https://github.com/WordPress/gutenberg/issues/34349

aaronrobertshaw commented:
> Given that there are no presets for radius, etc, so the custom prefix is spurious.

As André alluded to, these properties were added to be consistent with all the others given they had the `custom` prefix regardless of presets. Given what was there, I interpreted the `custom` prefix to mean that the theme was allowing users to set their own "custom" values as opposed to whatever was being set by the theme itself.

I'm happy to remove those prefixes in the v2 theme.json version outlined in https://github.com/WordPress/gutenberg/issues/34349. Sounds like it will make things more readable, less typing etc 👍 

> Also, let's disable border style for this one initially.

I'll bring this PR up-to-date, omit the border style support and give it all another test today.

oandregal commented:
Aaron: not for this PR, but I wonder if we could remove the `custom` prefix already from the properties that didn't land in WordPress 5.8. As far as [I see](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/), these are:

- border: all
- typography: customFontStyle, customFontWeight, customTextDecorations, customTextTransforms

It'd be best to do this as soon as possible in this cycle so themes have time to adapt before we start porting things for WordPress 5.9.

aaronrobertshaw commented:
I'm happy to put together a separate PR/s removing those `custom` prefixes in the next couple of days.

To confirm, only some supports landed in 5.8 and this didn't reflect the order they were merged in Gutenberg? I'm presuming because they are experimental features?

For the record, the proposed [height](https://github.com/WordPress/gutenberg/pull/32499) and [width](https://github.com/WordPress/gutenberg/pull/32502) support PRs have just had the prefix removed from them today.

glendaviesnz commented:
This tested well for me. Just one thing to note is that we probably want to think how the captions work in the gallery as adding a border to an image with the new Gallery innerBlocks format doesn't work with the current default caption setup:
<img width="217" alt="Screen Shot 2021-08-31 at 5 08 30 PM" src="https://user-images.githubusercontent.com/3629020/131445604-8e507514-cf30-4d23-9953-bf10d10ee39c.png">
but no change needed for this PR as new Gallery still under experimental flag so time to work on this separately, and gallery rather than image related. I have added a separate issue at https://github.com/WordPress/gutenberg/issues/34403

aaronrobertshaw commented:
> Just one thing to note is that we probably want to think how the captions work in the gallery as adding a border to an image with the new Gallery innerBlocks format doesn't work with the current default caption setup

Good point. I'll have to do some additional exploration on that front. 

If captions are being overlayed then a border around the `<figure>` element might work, if they aren't then the border would need to be on the `<img>` itself. Would potentially switching target elements cause a block validation/deprecation issue?

Applying the styles to different elements could be possible when applying individual blocks' styles so long as we can determine where they should be applied. For the theme.json styles, we are already having to skip the serialization of border styles so likely need to [extend the elements API](https://github.com/WordPress/gutenberg/blob/2b437fef3dd67863fdaf7af0e59b08e9fc678c15/lib/class-wp-theme-json-gutenberg.php#L242) to allow `img` elements. Between regular definition of styles in the theme.json and using the elements API within it, we might be able to cover both bases.

aaronrobertshaw committed:
Remove unneeded style now block support applied to img

aaronrobertshaw commented:
A PR addressing the removal of the `custom` prefix from properties that don't need it in the theme.json is up: https://github.com/WordPress/gutenberg/pull/34485

aaronrobertshaw committed:
Adjust caption styles for images in a gallery

aaronrobertshaw commented:
I've rebased this PR however it will still need further refinement. 

Along with the rebase I added a custom CSS class and tweaked styles to improve the display of image captions within the gallery when the image has a border. The gallery currently forces the caption beneath the image when the rounded block style is applied. For the moment, the same approach is used when a custom border is set.

<img width="724" alt="Screen Shot 2021-11-04 at 4 58 29 pm" src="https://user-images.githubusercontent.com/60436221/140272470-249d3cdf-952c-4eac-9c6f-45a1a99a3a73.png">

ramonjd commented:
> we are already having to skip the serialization of border styles so likely need to extend the elements API to allow img elements.

Noting as another block that might benefit from the outcomes in 
https://github.com/WordPress/gutenberg/issues/33255

aaronrobertshaw committed:
Fix typo

aaronrobertshaw committed:
Remove unneeded fallback border styles

aaronrobertshaw commented:
This PR has been rebased again.

After the recent switch of the border block support to use the `ToolsPanel`, the fallback styles added here [have been removed](https://github.com/WordPress/gutenberg/pull/31366/commits/231a8ec4687c13b07ca004b8b1b2fc8afe85deb5). 

In https://github.com/WordPress/gutenberg/pull/34061 the general decision was to make border width and color default controls for any block opting into it. I've also updated this PR to follow that. 

If the approach taken for handling captions on gallery images with borders is acceptable, this should be ok for another review.

ramonjd commented:
Hope you don't mind. I rebased on trunk to resolve conflict from https://github.com/WordPress/gutenberg/commit/ee07bbf5081ace0597d7558fc37de066257024a2

ramonjd commented:
This is still working well for me. 🎉 

Tried images and galleries, applying all border properties. The frontend reflects the editor. 

<img width="400" alt="Screen Shot 2021-12-15 at 9 58 14 am" src="https://user-images.githubusercontent.com/6458278/146094074-a95d9e30-7337-4c06-87dd-3ba328b1ca90.png">

Here's how it looks in Global Styles, but I understand there are ongoing discussions about how to target inner elements for  the Image and other blocks.

<img width="400" alt="Screen Shot 2021-12-15 at 10 00 31 am" src="https://user-images.githubusercontent.com/6458278/146094188-0c924416-154b-465a-af59-dcdbfd2cc285.png">

When testing cropping, I noticed that an image which has the rounded block style loses the border styles:

![Kapture 2021-12-15 at 10 06 18](https://user-images.githubusercontent.com/6458278/146093557-95d82f80-f3c1-47f5-b901-21ae89092d0f.gif)

I was testing on TwentyTwentyTwo and TT1 Blocks.

paaljoachim commented:
I downloaded the Gutenberg PR build.
Tested with
Chrome
OSX
Twenty Twenty Two

I was expecting this now to work natively without having to adjust anything in the theme.json file.
As I did not add any adjustments to the json file the result was that I noticed this:

![Screenshot 2021-12-15 at 01 21 54](https://user-images.githubusercontent.com/5323259/146100212-f2de4708-ee7a-4c8e-aec0-8f4618d47b12.jpg)

The greyed out options should be made available natively. It is better to leave things open and if the developer wants to turn various settings off they can do so through the json file.

aaronrobertshaw commented:
Thanks for testing @paaljoachim 👍 

> I was expecting this now to work natively without having to adjust anything in the theme.json file.

The TwentyTwentyTwo theme now sets `"appearanceTools": true` in its theme.json file. The [`appearanceTools` flag ](https://github.com/WordPress/gutenberg/pull/36646) turns on the border support in theme.json.

The test setup instructions here pre-date that addition to TwentyTwentyTwo and are still applicable if you are using different themes.

> The greyed out options should be made available natively. It is better to leave things open and if the developer wants to turn various settings off they can do so through the json file.

What you are seeing in your screenshot with the greyed out options is because [border color, style, and width are set as default controls ](https://github.com/WordPress/gutenberg/pull/31366/files#diff-4e637f96eb9e465696d35d73a41e6a694c7718f15fb805b49d49931262dfc53aR91-R95)within the Image block's border panel and have not yet been customised. You can actually see that those controls are available behind the dropdown in the screenshot.

With regards to the developer wishing to turn things off, if this was desired for TwentyTwentyTwo, an override for the `appearanceTools` flag can still be added. See https://github.com/WordPress/gutenberg/pull/36646 for more details on how.

glendaviesnz commented:
> The gallery currently forces the caption beneath the image when the rounded block style is applied. For the moment, the same approach is used when a custom border is set.

This tested well for me and seems like the best approach. It matches in well with how captions are handled when the radius is changed.

aaronrobertshaw committed:
Try improving application of custom border styles

aaronrobertshaw commented:
This PR has been rebased onto https://github.com/WordPress/gutenberg/pull/37770 so that it can benefit from the proposed new border controls and support for individual side borders. It has also been tweaked to provide a slightly better UX when resizing the image via drag handles while a custom border is present.

It still needs some additional testing and for #37770 to land before it will be ready for final review.

aaronrobertshaw committed:
Remove unneeded resizablebox style

aaronrobertshaw committed:
Tweak application of border-box sizing

aaronrobertshaw committed:
Remove border radius from align wrapper deprecation fixture

aaronrobertshaw commented on test/integration/fixtures/blocks/core__image__deprecated-align-wrapper.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:image {"align":"left","id":13,"width":358,"height":164,"sizeSlug":"large","linkDestination":"none","style":{"border":{"radius":"10px"}}} -->
-<div class="wp-block-image" style="border-radius:10px"><figure class="alignleft size-large is-resized"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==" alt="" class="wp-image-13" width="358" height="164"/></figure></div>
+<!-- wp:image {"align":"left","id":13,"width":358,"height":164,"sizeSlug":"large","linkDestination":"none"} -->
```
I've removed the border-radius style from this fixture as it wasn't related to the align wrapper deprecation. It's now covered by a dedicated deprecation required by this PR. See: [test/integration/fixtures/blocks/core__image__deprecated-border-radius-5.html](https://github.com/WordPress/gutenberg/pull/31366/files#diff-92e518972d78f9d6d1e174ff6a2331d7eb200e62aafedbd7ba243ad0248339b0)

aaronrobertshaw commented on packages/block-library/src/image/save.js:
```diff
@@ -27,18 +31,26 @@ export default function save( { attributes } ) {
 		[ `align${ align }` ]: align,
 		[ `size-${ sizeSlug }` ]: sizeSlug,
 		'is-resized': width || height,
+		'has-custom-border':
+			!! borderProps.className || ! isEmpty( borderProps.style ),
```
This new `has-custom-border` class was required to be able to style image captions appropriately when in a Gallery block given border support now needs to be applied directly on the internal `img` element.

aaronrobertshaw reviewed (commented):
This PR has now been brought up-to-date again and should be ready for another review.

beafialho commented:
I opened up (and ended up closing) an [issue](https://github.com/WordPress/gutenberg/issues/41229) to add borders to images not knowing this was already in progress, but just wanted to chime in and say how useful this can be in the context of style variations.

aaronrobertshaw committed:
Limit the damage done by theme.json and global styles borders

aaronrobertshaw committed:
Add img to Elements API and leverage for image block

aaronrobertshaw commented:
This is currently blocked by issues preventing the proper styling of image blocks via theme.json and Global Styles. 

In https://github.com/WordPress/gutenberg/pull/31366/commits/fa0b753979bbe2bd88ab1cacb3951afc9a2e7202, I've trialled adding `img` to the Elements API. This resolves most of the issues around theme.json styling however comes at the cost of requiring the element's selector to include `.wp-block-image__crop-area` to make the editor UX satisfactory. This means that same selector is included in the frontend CSS but that is a small price to pay until we can remove it when merged global styles are accessible from the block editor store.

Another downside of this approach is that Global Styles still applies image borders to the block's wrapper and not the `img` element itself.

An alternative approach might be to leverage `__experimentalSelector` however there is a bug currently preventing much exploration on that front (https://github.com/WordPress/gutenberg/issues/41296).

aaronrobertshaw committed:
Add img element to api constants

aaronrobertshaw committed:
Disable border UI in Global Styles when skipping serialization of support

aaronrobertshaw commented:
Given the multiple moving parts to supporting borders, e.g. theme.json, global styles, and individual block support, I thought I'd take another run at detailing the current state of this PR, its issues, and steps taken to mitigate those.

## What's the problem?


When we apply borders to an Image block, we're really wishing to apply borders to the block's inner `img` element. 

If border styles are applied to the block's wrapper, i.e. its root `figure` element, we'll be applying a border around not just the image but also the caption. 

<img width="400" alt="Screen Shot 2022-06-06 at 6 02 20 pm" src="https://user-images.githubusercontent.com/60436221/172120736-8c4a3cc9-880e-4e02-bb88-571085cdcc3b.png">

### Targeting the img element

We can skip the serialization of border styles on the Image block's wrapper (see `__experimentalSkipSerialization` ) and manually apply styles on the inner `img` element. This now gives us the ability to configure borders for individual blocks within the editor.

### The catch

This is where we start to get into muddier waters. Theme.json and Global Styles will generate their border styles for the block's selector e.g. `.wp-block-image`. This is effectively adding the border styles at the block's wrapper level again, giving us the same problem of the border wrapping both the inner `img` and `figcaption` elements.

### Theme.json

One option explored was to update the `__experimentalSelector` for the Image block with something to target the inner image e.g. `.wp-block-image > img`. This would affect more than just the border styles so wouldn't be ideal either. 

As it turned out, there is also a [bug](https://github.com/WordPress/gutenberg/issues/41296) preventing blocks outside of core from updating the `__experimentalSelector`.

The only approach I found to allow for applying borders to the inner image was to extend the Elements API for the `img` element. This would allow a theme author to apply a border via their theme.json and something similar to below:

```json
{
	"styles": {
		"core/image": {
			"elements": {
				"img": {
					"border": {
						"width": "15px",
						"style": "dashed",
						"color": "blue"
					}
				}
			}
		}
	}
}
```

At the cost of being nested deeper and a little different than most blocks, this gets us to the point where a theme author can style image blocks generally via theme.json and a user can style borders for individual blocks in the editor. 

The missing piece is now Global Styles in the site editor.

### Global Styles

Until we can apply border styles directly to CSS selectors of a block's choosing or target elements from the Elements API via the Global Styles UI, we will run into problems where a block support's styles are applied to the wrong element. Obviously, this isn't ideal.

This issue generally occurs when a block support's serialization is skipped. To prevent the issue in Global Styles, I've updated the border UI to only be available in block's that do not skip serialization.

In a nutshell, users will not be able to configure borders for images via Global Styles.


## Summary

1. Image border support works for individual blocks by skipping serialization
2. Image borders can be applied via theme.json through the Elements API and targeting `img`
3. The user **cannot** set generic image borders via Global Styles as the UI is disabled due to the generated styles applying incorrectly to the block's wrapper.

If having theme authors style image borders via the `img` element in theme.json and not allowing image border configuration via Global Styles is acceptable, I believe we can move forward with this PR and iterate on solutions to the issues raised above separately.

Does anyone have any fresh ideas or objections?

aaronrobertshaw commented:
## Duotone + Borders

Another minor issue is when a border is applied along with a duotone filter to an image. The filter will also apply to the image's border.

https://user-images.githubusercontent.com/60436221/172316154-ef315699-8f1d-44e6-8a9f-8ccc8628049f.mp4

aaronrobertshaw committed:
Revert "Disable border UI in Global Styles when skipping serialization of support"

This reverts commit 22af3f553958ac8b0dd2013c3c40e617bcd340c2.

aaronrobertshaw committed:
Revert "Add img to Elements API and leverage for image block"

This reverts commit fa0b753979bbe2bd88ab1cacb3951afc9a2e7202.

aaronrobertshaw committed:
Update image block to leverage feature level selectors for border styles

aaronrobertshaw committed:
Fix rebase issue after revert of img element

aaronrobertshaw commented:
Now https://github.com/WordPress/gutenberg/issues/41296 has been resolved, it opens up the possibility of adding individual block support feature level CSS selectors. This is being trialed within https://github.com/WordPress/gutenberg/pull/42087 and as such this PR has been updated to depend on that and now use these "feature level selectors" to apply theme.json and global styles borders to image blocks.

The result is now:
- Theme authors can correctly apply borders to the inner `img` element of image blocks without the need for an `img` option in the Elements API
- Users can actually configure image border styles within the Global Styles UI and have them correctly applied to an image block's inner `img` element. (This was the missing piece using the Elements API approach)
- Individual blocks can have their borders configured and applied appropriately, even while cropping.

Assuming the approach in https://github.com/WordPress/gutenberg/pull/42087 is acceptable and lands, this should be ready for a final review soon after.

aaronrobertshaw commented:
Now https://github.com/WordPress/gutenberg/issues/41296 has been resolved, it opens up the possibility of adding individual block support feature level CSS selectors. This is being trialed within https://github.com/WordPress/gutenberg/pull/42087 and as such this PR has been updated to depend on that and now use these "feature level selectors" to apply theme.json and global styles borders to image blocks.

The result is now:
- Theme authors can correctly apply borders to the inner `img` element of image blocks without the need for an `img` option in the Elements API
- Users can actually configure image border styles within the Global Styles UI and have them correctly applied to an image block's inner `img` element. (This was the missing piece using the Elements API approach)
- Individual blocks can have their borders configured and applied appropriately, even while cropping.

Assuming the approach in https://github.com/WordPress/gutenberg/pull/42087 is acceptable and lands, this should be ready for a final review soon after.

aaronrobertshaw commented:
Now https://github.com/WordPress/gutenberg/issues/41296 has been resolved, it opens up the possibility of adding individual block support feature level CSS selectors. This is being trialed within https://github.com/WordPress/gutenberg/pull/42087 and as such this PR has been updated to depend on that and now use these "feature level selectors" to apply theme.json and global styles borders to image blocks.

The result is now:
- Theme authors can correctly apply borders to the inner `img` element of image blocks without the need for an `img` option in the Elements API
- Users can actually configure image border styles within the Global Styles UI and have them correctly applied to an image block's inner `img` element. (This was the missing piece using the Elements API approach)
- Individual blocks can have their borders configured and applied appropriately, even while cropping.

Assuming the approach in https://github.com/WordPress/gutenberg/pull/42087 is acceptable and lands, this should be ready for a final review soon after.

aaronrobertshaw committed:
Override clearing of img borders in wp-admin by core

paaljoachim commented:
Hi @aaronrobertshaw 
Let us know here in the PR (when ready) when to test and how to test.
Thanks!

aaronrobertshaw commented:
Thanks for checking in on this one @paaljoachim 👍 

You can test this PR now via the test instructions in the PR description. If you have any troubles with this please let me know.

As I noted in my [previous comment](https://github.com/WordPress/gutenberg/pull/31366#issuecomment-1172110651), this PR depends upon https://github.com/WordPress/gutenberg/pull/42087 landing first. https://github.com/WordPress/gutenberg/pull/42087 is required so that we can style images via Global Styles.

aaronrobertshaw commented:
@paaljoachim This PR is ready for further testing and hopefully a final review 🎉

ramonjd commented on packages/block-library/src/image/block.json:
```diff
@@ -84,9 +84,15 @@
+			"color": true,
 			"radius": true,
+			"width": true,
+			"__experimentalSelector": "img, .wp-block-image__crop-area",
```
🍺

ramonjd reviewed (approved):
I've taken this for a run in the editor and it works as advertised for me!

✅ In the editor the block border css declarations appear on the `img` and `.wp-block-image__crop-area` elements. The border persists during and after cropping.
✅ Image borders for images and gallery images look good on the frontend
✅ Creating huge border widths in a gallery does not collapse columns
✅ `styles.blocks['core/image'].border` styles are applied appropriately 
✅ Global styles updated in the site editor are applied appropriately (and taken precedence over theme.json)

andrewserong reviewed (approved):
Great work getting this to the finish line @aaronrobertshaw! Tested with a few different block and classic themes, and all working nicely 👍

✅ Crop tools are working nicely in the editor with and without a custom border
✅ Custom border color is working nicely in the editor and site editor
✅ Image border in global styles works well
✅ Deprecation changes look good to me visually

The only visual issue I encountered was with a caption on an image in a gallery block, where ~that image has a border and radius~ in global styles the Image block is set to have a border and radius. The linear gradient and absolute positioning of the figcaption mean that it extends over the area of the border radius:

<img width="593" alt="image" src="https://user-images.githubusercontent.com/14988353/179154541-80716bf1-dfd2-4a30-9cf1-856e95e984e7.png">

Given the effort already to ensure that the border (and its radius) does not affect the `figcaption`, and that this is handled for individual blocks with radii, I think this isn't worth attempting to fix in this PR, particularly when separately there's work and recent PRs looking at customising the figcaption.

In the "tiny nits" department, elsewhere there's efforts to remove lodash as a dependency, so in a follow-up at some point, it might be worth us seeing if we can swap out the `isEmpty` calls for another approach — given the iterations on getting to this stage with this PR, though, definitely not a blocker for landing this one.

So, all in all this LGTM! 🎉

aaronrobertshaw commented:
I appreciate the reviews @ramonjd and @andrewserong, cheers 🍻 

> The only visual issue I encountered was with a caption on an image in a gallery block, where that image has a border and radius in global styles the Image block is set to have a border and radius. The linear gradient and absolute positioning of the figcaption mean that it extends over the area of the border-radius:

Thanks for raising this edge case. 

When borders are applied individually to blocks, we can add an additional CSS class to them that allows us to adjust the positioning of the caption. When borders are configured via Global Styles they are targeted via a simple CSS selector and the block won't know whether or not the caption should be repositioned.

There could be a few options to explore in follow-ups. 
- As you note, if the caption becomes an inner block to the Image block, then it could provide styling options. 
- Additionally, the Gallery block could offer an option, or [Block Style](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-styles/), to toggle all captions to overlay the image or not. This might provide better visual results in terms of consistency and only require a single update rather than across multiple blocks.

> In the "tiny nits" department, elsewhere there's efforts to remove lodash as a dependency, so in a follow-up at some point, it might be worth us seeing if we can swap out the isEmpty calls for another approach

Agreed. This is a side effect of just how old and long-running this PR has been. I'll happily address this in a separate follow-up.

paaljoachim commented:
Hi @aaronrobertshaw 

Thank you very much for working on this feature Aaron! It is greatly appreciated!
I look forward to one day be able to save adjustments as a style and use it for all images.

Testing.
Lock and unlock icons. This is how I see it.
Lock icon = means all sides have the same width and color.
Unlock icon = means that width and/or color is not the same for all sides.

Clicking the Lock/Unlock icon will open or close the individual sides attributes/properties.

Example.
Here is a screenshot of Border where individual sides have their own property (likely wrong word to use) as in different color/width.
Radius is the same for all sides.

I would expect to see the Border icon show as unlocked because I have gone in and adjusted the properties of different sides.
But here we see that both icons are identical even though Border has individual side changes and Radius does not.

![Screenshot 2022-07-15 at 15 55 25](https://user-images.githubusercontent.com/5323259/179238134-9068ac40-167c-4280-85fe-885651f5348e.jpg)
